plugins/lookup/hashi_vault.py

```diff
@@ -318,6 +334,10 @@ def __init__(self, **kwargs):
             client_args['token'] = self.options.get('token')
 
         self.client = hvac.Client(**client_args)
+        # logout to prevent accidental use of inferred tokens
+        # https://github.com/ansible-collections/community.hashi_vault/issues/13
+        if 'token' not in client_args:
+            self.client.logout()
```

Comment:
I think this requires I changelog fragment.

Reply:
@felixfontein I don't think it does but certainly willing to hear you out! 

This doesn't change current behavior, because every auth method (except token), sets the token in the client after this point (all of them do it implicitly except JWT, for that one we do it explicitly in this plugin; that's an HVAC inconsistency).

For token auth, we set that before creating the client, so we don't call logout.

Calling `.logout()` here is to ensure that the token is clear before we continue to the auth methods, because when a token is not supplied during client creation, it tries to find one (in env or in token file). In all normal cases that doesn't matter: the auth  method will overwrite it.

So what this really does is prevent a developer from accidentally not ensuring a token is set in a new auth method (or when modifying an existing one). Before this change, what would happen is that the client still had its inferred token, so it would use it in further operations, which would give a confusing result. 

That's especially true in our CI where a root token with full privileges is available. This was surfaced in the PR to add JWT auth, when all the tests were passing except the one that tried to access a secret it was not supposed to have access to (expecting access denied), and inexplicably it was able to access the secret.

It was seemingly so weird we were considering whether it was a bug in vault and whether to remove the test (which would have been a disaster because in truth the auth method was never actually setting the client with the token it received from doing JWT auth).

--- 

So in short, all existing auth methods would never trigger this, this change is to prevent future implementations from passing a bunch of tests by accident because of re-used tokens.